src/coreclr/jit/importercalls.cpp

```diff
@@ -2759,6 +2762,81 @@ GenTree* Compiler::impIntrinsic(GenTree*                newobjThis,
                 break;
             }
 
+            case NI_System_Runtime_CompilerServices_RuntimeHelpers_IsReference:
+            {
+                assert(sig->sigInst.methInstCount == 1);
+                CORINFO_CLASS_HANDLE hClass = sig->sigInst.methInst[0];
+                retNode = gtNewIconNode(eeIsValueClass(hClass) ? 0 : 1);
+                break;
+            }
+
+            case NI_System_Runtime_CompilerServices_RuntimeHelpers_IsReferenceOrContainsReferences:
+            {
+                assert(sig->sigInst.methInstCount == 1);
+                CORINFO_CLASS_HANDLE hClass = sig->sigInst.methInst[0];
+                retNode = gtNewIconNode(!eeIsValueClass(hClass) || ((info.compCompHnd->getClassAttribs(hClass) & CORINFO_FLG_CONTAINS_GC_PTR) != 0) ? 1 : 0);
+                break;
+            }
+
+            case NI_System_Runtime_CompilerServices_RuntimeHelpers_IsBitwiseEquatable:
+            {
+                assert(sig->sigInst.methInstCount == 1);
+                CORINFO_CLASS_HANDLE hClass = sig->sigInst.methInst[0];
+                retNode = gtNewIconNode(info.compCompHnd->isBitwiseEquatable(hClass) ? 1 : 0);
+                break;
+            }
+
+            case NI_System_Runtime_CompilerServices_RuntimeHelpers_GetMethodTable:
+            {
+                retNode = gtNewMethodTableLookup(impPopStack().val);
+                break;
+            }
+
+            case NI_System_Runtime_CompilerServices_RuntimeHelpers_EnumEquals:
+            {
+                assert(sig->sigInst.methInstCount == 1);
+                CORINFO_CLASS_HANDLE hClass  = sig->sigInst.methInst[0];
+                CorInfoType          jitType = info.compCompHnd->asCorInfoType(hClass);
+
+                if (varTypeIsFloating(JitType2PreciseVarType(jitType)))
+                {
+                    return nullptr;
+                }
+
+                GenTree* op2 = impPopStack().val;
+                GenTree* op1 = impPopStack().val;
+                retNode = impImportCompare(op1, op2, GT_EQ, false);
+                break;
+            }
+
+            case NI_System_Runtime_CompilerServices_RuntimeHelpers_EnumCompareTo:
+            {
+                assert(sig->sigInst.methInstCount == 1);
+
+                CORINFO_CLASS_HANDLE hClass  = sig->sigInst.methInst[0];
+                CorInfoType          jitType = info.compCompHnd->asCorInfoType(hClass);
+                var_types            type    = JitType2PreciseVarType(jitType);
+
+                if (varTypeIsFloating(type))
+                {
+                    return nullptr;
+                }
+
+                GenTree* op2 = impPopStack().val;
+                GenTree* op1 = impPopStack().val;
+                bool     uns = varTypeIsUnsigned(type);
+
+                GenTree* op1Clone;
+                GenTree* op2Clone;
+                op1 = impCloneExpr(op1, &op1Clone, CHECK_SPILL_ALL, nullptr DEBUGARG("EnumCompareTo arg1"));
```

Comment:
This is less efficient that the existing implementation for 8-bit and 16-bit enums.

Also, this may may better be marked as `mustExpand` so that there are no observable behavior differences between the expanded and non-expanded version.

Reply:
The existing implementation that just called CompareTo on the underlying type feels more robust, less duplicative.

Reply:
> This is less efficient that the existing implementation for 8-bit and 16-bit enums.

Fixed.

> Also, this may be better marked as `mustExpand` so that there are no observable behavior differences between the expanded and non-expanded version.

It's marked as `betterExpand` now, which means that it'll expand unless in debug AFAIR. I don't think we should have any behaviour differences between the two paths.

> The existing implementation that just called CompareTo on the underlying type feels more robust, less duplicative.

But it also relies on that hardcoded table of tables for ILs for selected types there and it relies more on the inliner. This logic resembles the Mono one where it's expanded directly in the intrinsic too. We could potentially change this to `PrimiviteCompareTo` and use it for all primitives in their `CompareTo`s or we could just remove the intrinsic and rewrite the code in managed with the `Enum.GetUnderlyingType` intrinsic instead (but that could potentially run into inlining issues due to IL size).

Reply:
> It's marked as betterExpand now, which means that it'll expand unless in debug AFAIR. 

It means that the debug code will allocate vs. release code will not. The performance difference between debug and release is going to be order of magnitude more than what it is today. Performance regressions like this are not acceptable for debug code. It is not unusual to see issues filled on large debug code performance regressions like this by our customers.

> relies on that hardcoded table of tables for ILs for selected types there 

It is not hard to fix the code so that it works for all enum underlying types. 

> it relies more on the inliner

The method being inlined is very small. There are thousands of places in this repo where we have small methods to make the code easier to maintain and depend on them being inlined without an issue. Why should this one be different? Enum comparers are not that common for inliner avoidance to make difference.

> we could just remove the intrinsic and rewrite the code in managed with the Enum.GetUnderlyingType intrinsic instead (but that could potentially run into inlining issues due to IL size).

Right, this approach would not work well.